cmake/ccf.cmake

```diff
@@ -0,0 +1,181 @@
+    target_link_libraries(${name} PRIVATE ${COVERAGE_LINK})
+  endif()
+endfunction()
+
+function(use_client_mbedtls name)
+  target_include_directories(${name} PRIVATE ${CLIENT_MBEDTLS_INCLUDE_DIR})
+  target_link_libraries(${name} PRIVATE ${CLIENT_MBEDTLS_LIBRARIES})
+endfunction()
+
+function(use_oe_mbedtls name)
+  target_link_libraries(${name} PRIVATE
+    openenclave::oeenclave
+    openenclave::oelibcxx
+  )
+endfunction()
+
+if(NOT CCF_GENERATED_DIR)
+  set(CCF_GENERATED_DIR ${CCF_DIR}/generated)
+endif()
+
+add_custom_target(flatbuffers ALL
+  DEPENDS ${CCF_GENERATED_DIR}/frame_generated.h
+)
+
+# Enclave library wrapper
+function(add_enclave_lib name)
+
+  cmake_parse_arguments(PARSE_ARGV 1 PARSED_ARGS
+    ""
+    ""
+    "SRCS;INCLUDE_DIRS;LINK_LIBS"
+  )
+
+  if("sgx" IN_LIST TARGET)
+    add_library(${name} SHARED
+      ${ENCLAVE_FILES}
+      ${PARSED_ARGS_SRCS}
+      ${CCF_GENERATED_DIR}/ccf_t.cpp
+    )
+
+    target_compile_definitions(${name} PRIVATE
+      INSIDE_ENCLAVE
+      _LIBCPP_HAS_THREAD_API_PTHREAD
+    )
+    target_compile_options(${name} PRIVATE
+      -nostdinc
+      -nostdinc++
+      -U__linux__
+    )
+    target_include_directories(${name} SYSTEM PRIVATE
+      ${PARSED_ARGS_INCLUDE_DIRS}
+      ${EVERCRYPT_INC}
+      ${CMAKE_CURRENT_BINARY_DIR}
+      ${QUICKJS_INC}
+    )
+    add_dependencies(${name} flatbuffers)
+
+    if (PBFT)
+      target_link_libraries(${name} PRIVATE
+        -Wl,--allow-multiple-definition #TODO(#important): This is unfortunate
+        libbyz.enclave
+      )
+    endif()
+    target_link_libraries(${name} PRIVATE
+      -nostdlib -nodefaultlibs -nostartfiles
+      -Wl,--no-undefined
+      -Wl,-Bstatic,-Bsymbolic,--export-dynamic,-pie
+      quickjs.enclave
+      -lgcc
+      ${PARSED_ARGS_LINK_LIBS}
+      openenclave::oeenclave
+      openenclave::oelibcxx
+      ${ENCLAVE_LIBS}
+      http_parser.enclave
+    )
+    set_property(TARGET ${name} PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+    enable_quote_code(${name})
+  endif()
+
+  if("virtual" IN_LIST TARGET)
+    ## Build a virtual enclave, loaded as a shared library without OE
+    set(virt_name ${name}.virtual)
+    add_library(${virt_name} SHARED
+      ${ENCLAVE_FILES}
+      ${PARSED_ARGS_SRCS}
+      ${CCF_GENERATED_DIR}/ccf_t.cpp
+    )
+    target_compile_definitions(${virt_name} PRIVATE
+      INSIDE_ENCLAVE
+      VIRTUAL_ENCLAVE
+    )
+    target_compile_options(${virt_name} PRIVATE
+      -stdlib=libc++)
+    target_include_directories(${virt_name} SYSTEM PRIVATE
+      ${PARSED_ARGS_INCLUDE_DIRS}
+      ${CCFCRYPTO_INC}
+      ${EVERCRYPT_INC}
+      ${CMAKE_CURRENT_BINARY_DIR}
+      ${QUICKJS_INC}
+      ${OE_INCLUDEDIR} # Virtual libraries don't link against OE, but do share includes
+    )
+    add_dependencies(${virt_name} flatbuffers)
+
+    if (PBFT)
+      target_link_libraries(${virt_name} PRIVATE
+        -Wl,--allow-multiple-definition #TODO(#important): This is unfortunate
```

Comment:
Not really part of this PR but since we removed the `libbyz/keypair.h` I think this doesn't have to be unfortunate anymore (I think we can remove the `-Wl, --alow-multiple-definitions`)

Reply:
Nice - have removed this unpleasantness entirely.